cli/core/test/e2e/framework/plugin_lifecycle_operations.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package framework
+
+import (
+	"fmt"
+	"strings"
+
+	"encoding/json"
+
+	"github.com/pkg/errors"
+)
+
+// PluginBasicOps helps to perform the plugin command operations
+type PluginBasicOps interface {
+	// ListPlugins lists all plugins by running 'tanzu plugin list' command
+	ListPlugins() ([]PluginListInfo, error)
+}
+
+// PluginSourceOps helps 'plugin source' commands
+type PluginSourceOps interface {
+	// AddPluginDiscoverySource adds plugin discovery source
+	AddPluginDiscoverySource(discoveryOpts *DiscoveryOptions) error
+}
+
+// PluginLifecycleOps helps to perform the plugin and its sub-commands lifecycle operations
+type PluginLifecycleOps interface {
+	PluginBasicOps
+	PluginSourceOps
+}
+
+type DiscoveryOptions struct {
+	Name       string
+	SourceType string
+	URI        string
+}
+
+type pluginLifecycleOps struct {
+	cmdExe CmdOps
+	PluginLifecycleOps
+}
+
+func NewPluginLifecycleOps() PluginLifecycleOps {
+	return &pluginLifecycleOps{
+		cmdExe: NewCmdOps(),
+	}
+}
+
+func (po *pluginLifecycleOps) AddPluginDiscoverySource(discoveryOpts *DiscoveryOptions) error {
+	addCmd := fmt.Sprintf(AddPluginSource, discoveryOpts.Name, discoveryOpts.SourceType, discoveryOpts.URI)
+	_, _, err := po.cmdExe.Exec(addCmd)
+	return err
+}
+
+func (po *pluginLifecycleOps) ListPlugins() ([]PluginListInfo, error) {
+	out, _, err := po.cmdExe.Exec(ListPluginsCmd)
+	if err != nil {
+		return nil, err
+	}
+	var list []PluginListInfo
+	outStr := out.String()
+	start := strings.LastIndex(outStr, "[\n")
+	end := strings.LastIndex(outStr, "]")
+	jsonData := outStr[start : end+1]
```

Comment:
Why do we need these lines that try to remove `[` and `]` from JSON output?
As I mentioned in my previous example https://go.dev/play/p/EApTwngSCKx, I am not sure we need to remove this.

Reply:
the CLI output has extra string apart from the json data, eg:

`!  unable to list plugin from discovery 'oci_local': error while processing package: failed to get resource files from discovery: error downloading bundle: Checking if image is bundle: Fetching image: GET http://localhost:5001/v2/tkg/tanzu_core/tanzu-cli-plugins/k8s/dummy___/manifests/latest: unexpected status code 404 Not Found: 404 page not found

[
  {
    "description": "its dummy plugin",
    "discovery": "oci-a1",
    "name": "dummy",
    "scope": "Standalone",
    "status": "not installed",
    "version": ""
  }]`
above code considers json data between and include '[' and ']'

Reply:
`!  unable to list plugin from discovery 'oci_local': error while processing package: failed to get resource files from discovery: error downloading bundle: Checking if image is bundle: Fetching image: GET http://localhost:5001/v2/tkg/tanzu_core/tanzu-cli-plugins/k8s/dummy___/manifests/latest: unexpected status code 404 Not Found: 404 page not found
`
This log-line should not be getting printed to `stdout` and it should go to `stderr`. 
While reading this output of a command to process the output we should just read from the `stdout` and ignore `stderr`. 

Considering this, if the above log-line is getting printed to `stdout` than it is a bug and we should file that bug as part of this repository. Along with that bug fix, we should change this logic as well as this is a hack that we are doing which is not a good practice.

@chandrareddyp can you please help with filling the issue?

Reply:
Also regarding `oci_local`, are we adding this discovery source to our config file? I am trying to understand where this error comes from.

Reply:
I have created a bug to track it, and the above logic will be removed after fixing the bug. 
The `oci_local` is one of the discovery source with incorrect URL that's why its returning error.

Reply:
As discussed, i have tested the code without any invalid plugin URL sources, its working fine, so i have removed the json output parsing, its working fine. thanks.